### PR TITLE
[ENH] Rate matrix Hamiltonian Monte Carlo

### PR DESCRIPTION
**Pull request by rmcgibbo**
This is not ready to merge yet.

### CODE REVIEW

**Comment by rmcgibbo** [comment]
I think this is ready to be merged.

**Comment by kyleabeauchamp** [comment]
Looks like scipy has released 0.15.1 that fixes the statsmodels issue now.  Not sure whether we should do anything about it.  Can we use `scipy != 0.15.0`?

File: msmbuilder/msm/_ratematrix_priors.pyx [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    if grad is not None:
+        for i in range(n):
+            grad[i] += (-exptheta[i] * (sumalpha / norm) + alpha[i] - sumtheta)
+
+    return logp
+
+
+def lexponential(const double[::1] theta, const double[::1] beta,
+                 double[::1] grad=None):
+    r"""Log probability of the expnential distribution. Reparameterized
```
Comment (kyleabeauchamp) [inline msmbuilder/msm/_ratematrix_priors.pyx]: exponential

File: msmbuilder/msm/bayes_ratematrix.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+        Concentration parameter for a Dirichlet prior on the equilibrium
+        distribution.
+    prior_beta : float > 0 or ndarray, shape=(len(theta_)-n_states_,)
+        Scale parameter for exponential prior on the symmetric rate matrix.
+    n_timescales : int, optional
+        Number of implied timescales to calculate.
+    use_sparse : bool, default=True
+        Constrain the rate matrix to the sparsity pattern induced by the MLE
+        during sampling.
+    sliding_window : bool, optional
```
Comment (kyleabeauchamp) [inline msmbuilder/msm/bayes_ratematrix.py]: Document the default here as it's one of the more important choices to make?

**Comment by kyleabeauchamp** [comment]
First pass through and it seems reasonable.  I haven't looked at the upstream hmc code yet.

**Pull request merged**